packages/components/src/utils/hooks/use-focus-outside.js

```diff
@@ -0,0 +1,211 @@
+/**
+ * External dependencies
+ */
+import { includes } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback, useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Input types which are classified as button types, for use in considering
+ * whether element is a (focus-normalized) button.
+ *
+ * @type {string[]}
+ */
+const INPUT_BUTTON_TYPES = [ 'button', 'submit' ];
+
+/**
+ * @typedef {HTMLButtonElement | HTMLLinkElement | HTMLInputElement} FocusNormalizedButton
+ */
+
+// Disable reason: Rule doesn't support predicate return types
+/* eslint-disable jsdoc/valid-types */
+/**
+ * Returns true if the given element is a button element subject to focus
+ * normalization, or false otherwise.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+ *
+ * @param {EventTarget} eventTarget The target from a mouse or touch event.
+ *
+ * @return {eventTarget is FocusNormalizedButton} Whether element is a button.
+ */
+function isFocusNormalizedButton( eventTarget ) {
+	if ( ! ( eventTarget instanceof window.HTMLElement ) ) {
+		return false;
+	}
+	switch ( eventTarget.nodeName ) {
+		case 'A':
+		case 'BUTTON':
+			return true;
+
+		case 'INPUT':
+			return includes(
+				INPUT_BUTTON_TYPES,
+				/** @type {HTMLInputElement} */ ( eventTarget ).type
+			);
+	}
+
+	return false;
+}
+/* eslint-enable jsdoc/valid-types */
+
+/**
+ * @typedef {import('react').SyntheticEvent} SyntheticEvent
+ */
+
+/**
+ * @callback EventCallback
+ * @param {SyntheticEvent} event input related event.
+ */
```

Comment:
This highlights one of the limitations of JSDoc compared to TypeScript. Default type arguments cannot be expressed, and we have to provide type parameters on the imported types. If we want to rely on default type arguments, we either cannot do a single import like `/** @typedef {import('react').SyntheticEvent} SyntheticEvent */` or we have to do imports for each of the default types. I'll try to clarify

```ts
// 👇 SyntheticEvent accepts type parameters, but now we can't pass them. This is "fixed" as `SyntheticEvent<Element, Event>`.
/**
 * @typedef {import('react').SyntheticEvent} SyntheticEvent
 */

// Now, although later we can reason about different event types we can't pass them into synthetic event 😞
/**
 * @callback EventCallback
 * @param {SyntheticEvent} event input related event.
 */

// If we want to wrap an imported type but keep (some of) the type parameters, we have to import here:
/**
 * @template {Event} E
 * @callback EventCallback
 * @param {SyntheticEvent<Element, E>} event input related event.
 * @return void
 */

// Now, for e.g. `handleFocusOutside` we can be very specific that it's a _focus_ event handler
/**
 * @typedef FocusOutsideReactElement
 * @property {EventCallback<FocusEvent>} handleFocusOutside callback for a focus outside event.
 */
```

This is a tradeoff, if we want an `EventCallback` for `Event`, we need to type it as `EventCallback<Event>` or include another typedef without a type parameter.

We don't need to consider any of these tradeoffs in TypeScript syntax, it's better:

```ts
import type { SyntheticEvent } from 'react'
type SyntheticEventHandler<E extends Event = Event> = (e: SyntheticEvent<Element, E>) => void;
interface HandleFocusOutside {
    handleFocusOutside: SyntheticEventHandler<FocusEvent>;
    handleAnyEvent: SyntheticEventHandler; // No type params, handles `Event` (for demonstration)
}
```

Reply:
Yeah, I noticed this when trying to alias `MutableRefObject` to reduce the line length of some of the documentation. That doesn't have a default so it can't be aliased at all.